File: Coding-Patterns/Coding-Patterns-Python/08_linked_lists/merge_k_lists.py
# LeetCode 23 - Merge k Sorted Lists
# Difficulty: Hard
# Pattern: Min-heap over list heads / divide-and-conquer pairwise merge
#
# Problem: Given k sorted linked lists, merge them into one sorted list.
#
# Key insight: the answer's next node is always the smallest of the k current
# heads. Both approaches below exploit that, they just differ in how they keep
# track of "smallest":
#
#   merge_k_lists          - heap of the k live heads.      O(N log k) / O(k)
#   merge_k_lists_divide   - merge lists pairwise in rounds. O(N log k) / O(log k)
#
# The naive alternative - merge list 1 into the accumulator, then list 2, ... -
# is O(N * k): the accumulator gets rescanned on every merge.
#
# Both reuse the input nodes rather than allocating new ones, so the input
# lists are consumed (their `next` pointers get rewired).

import heapq
from typing import List, Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def merge_k_lists(lists: List[Optional[ListNode]]) -> Optional[ListNode]:
    """Min-heap of the k live heads. Time: O(N log k)  Space: O(k)"""
    # ListNode is not comparable, and equal values would make heapq fall through
    # to comparing nodes. The monotonic `order` breaks every tie first.
    heap = []
    order = 0
    for node in lists:
        if node:                      # skip the empty lists the constraints allow
            heapq.heappush(heap, (node.val, order, node))
            order += 1

    dummy = ListNode(0)
    tail = dummy

    while heap:
        _, _, node = heapq.heappop(heap)
        tail.next = node
        tail = node

        if node.next:
            heapq.heappush(heap, (node.next.val, order, node.next))
            order += 1

    tail.next = None                  # cut the tail loose from its old list
    return dummy.next


def merge_k_lists_divide(lists: List[Optional[ListNode]]) -> Optional[ListNode]:
    """Pairwise merge, halving the list count each round.

    log k rounds, each touching all N nodes.
    Time: O(N log k)  Space: O(log k) recursion
    """
    if not lists:
        return None
    return _merge_range(lists, 0, len(lists) - 1)


def _merge_range(lists, left, right):
    if left == right:
        return lists[left]

    mid = (left + right) // 2
    return merge_two(_merge_range(lists, left, mid),
                     _merge_range(lists, mid + 1, right))


def merge_two(a: Optional[ListNode], b: Optional[ListNode]) -> Optional[ListNode]:
    """Standard two-pointer merge (see merge_two_sorted_lists.py)."""
    dummy = ListNode(0)
    curr = dummy

    while a and b:
        if a.val <= b.val:
            curr.next = a
            a = a.next
        else:
            curr.next = b
            b = b.next
        curr = curr.next

    curr.next = a or b
    return dummy.next


# ---- Helpers ----
def to_list(node):
    result = []
    while node:
        result.append(node.val)
        node = node.next
    return result

def from_list(vals):
    dummy = ListNode(0)
    curr = dummy
    for v in vals:
        curr.next = ListNode(v)
        curr = curr.next
    return dummy.next


# ---- Tests ----
if __name__ == "__main__":
    cases = [
        ([[1, 4, 5], [1, 3, 4], [2, 6]], [1, 1, 2, 3, 4, 4, 5, 6]),
        ([], []),                              # k == 0
        ([[]], []),                            # one empty list
        ([[], [], []], []),                    # all empty
        ([[], [1], []], [1]),                  # empties interleaved with content
        ([[1, 2, 3]], [1, 2, 3]),              # k == 1
        ([[2, 2], [2], [2, 2, 2]], [2] * 6),   # all-duplicate values (tie-break)
        ([[-10000, 0], [-5, 10000]], [-10000, -5, 0, 10000]),
    ]

    for lists, expected in cases:
        # Each solver consumes its input, so build the nodes fresh per solver.
        assert to_list(merge_k_lists([from_list(v) for v in lists])) == expected
        assert to_list(merge_k_lists_divide([from_list(v) for v in lists])) == expected

    # k at the constraint edge: 10^4 lists, 10^4 nodes total.
    wide = [from_list([i]) for i in range(10_000)]
    assert to_list(merge_k_lists(wide)) == list(range(10_000))

    print("All tests passed.")

File: Programs-List-Repository/MergeKLists.cs
/*
Merge k sorted linked lists and return it as one sorted list.

Time Complexity: O(N log k) - where N is the total number of nodes in all lists and k is the number of linked lists.
        Each node is inserted into the priority queue once, and the priority queue has a size of at most k, so each insertion takes O(log k) time.
space Complexity: O(k) - The priority queue can hold at most k nodes at any time, where k is the number of linked lists.
*/

namespace CodingPatterns.LinkedLists;

public class ListNode<T> {
    public T val;
    public ListNode<T> next;
    public ListNode(T val = default, ListNode<T> next = null) {
        this.val = val;
        this.next = next;
    }
}

public class MergeKLists<T> where T : IComparable<T> {
    public ListNode<T> Merge(ListNode<T>[] lists) {
        if (lists == null || lists.Length == 0) 
            return null;

        // Min-heap ordered by node value; tie-break with an insertion index
        // to keep comparisons stable (PriorityQueue requires a comparable key).
        var pq = new PriorityQueue<ListNode<T>, (T val, int order)>();

        int order = 0;
        foreach (var node in lists) {
            if (node != null) {
                pq.Enqueue(node, (node.val, order++));
            }
        }

        var dummy = new ListNode<T>();
        var tail = dummy;

        while (pq.Count > 0) {
            var node = pq.Dequeue();
            tail.next = node;
            tail = tail.next;

            if (node.next != null) {
                pq.Enqueue(node.next, (node.next.val, order++));
            }
        }

        tail.next = null;
        return dummy.next;
    }

    /*
    Recursive divide-and-conquer: pair up lists and merge each pair, halving
    the number of lists every round until only one remains.

    Time Complexity: O(N log k) - there are log k rounds of pairwise merges,
            and each round merges a total of N nodes across all pairs.
    Space Complexity: O(log k) - for the recursion call stack (plus O(1)
            extra space per merge, since MergeTwoLists is iterative and
            reuses existing nodes rather than allocating new ones).
    */
    public ListNode<T> MergeKListsRecursive(ListNode<T>[] lists) {
        if (lists == null || lists.Length == 0) return null;
        return Merge(lists, 0, lists.Length - 1);
    }

    private ListNode<T> Merge(ListNode<T>[] lists, int left, int right) {
        if (left == right) return lists[left];

        int mid = left + (right - left) / 2;
        var leftMerged = Merge(lists, left, mid);
        var rightMerged = Merge(lists, mid + 1, right);
        return MergeTwoLists(leftMerged, rightMerged);
    }

    private ListNode<T> MergeTwoLists(ListNode<T> a, ListNode<T> b) {
        var dummy = new ListNode<T>();
        var tail = dummy;

        while (a != null && b != null) {
            if (a.val.CompareTo(b.val) <= 0) {
                tail.next = a;
                a = a.next;
            } else {
                tail.next = b;
                b = b.next;
            }
            tail = tail.next;
        }

        tail.next = a ?? b;
        return dummy.next;
    }
}

// Run() cannot live on MergeKLists<T> itself: an open generic type has no
// invocable static entry point, so the demo gets its own closed host class.
public static class MergeKListsDemo
{
    public static void Main() {
        var solver = new MergeKLists<int>();

        // Each solver rewires the input nodes, so every run gets fresh lists.
        ListNode<int>[] Build(params int[][] values) =>
            values.Select(FromArray).ToArray();

        var inputs = new[] {
            new[] { new[] { 1, 4, 5 }, new[] { 1, 3, 4 }, new[] { 2, 6 } },  // the example
            Array.Empty<int[]>(),                                            // k == 0
            new[] { Array.Empty<int>() },                                    // one empty list
            new[] { Array.Empty<int>(), new[] { 1 }, Array.Empty<int>() },   // empties around content
            new[] { new[] { 2, 2 }, new[] { 2 }, new[] { 2, 2, 2 } },        // all ties
            new[] { new[] { -10000, 0 }, new[] { -5, 10000 } },              // value range edges
        };

        foreach (var values in inputs) {
            var heap = ToArray(solver.Merge(Build(values)));
            var divide = ToArray(solver.MergeKListsRecursive(Build(values)));
            Console.WriteLine($"[{string.Join(" ", heap)}]  heap == divide&conquer: {heap.SequenceEqual(divide)}");
        }
    }

    private static ListNode<int> FromArray(int[] values) {
        var dummy = new ListNode<int>();
        var tail = dummy;
        foreach (var v in values)
            tail = tail.next = new ListNode<int>(v);
        return dummy.next;
    }

    private static List<int> ToArray(ListNode<int> head) {
        var values = new List<int>();
        for (var node = head; node != null; node = node.next)
            values.Add(node.val);
        return values;
    }
}
